### ISSUE TITLE
Clippy fails with E0228, but build won't

### ISSUE THREAD

**Issue by 06393993** [rust-lang/rust-clippy#11417]
### Summary

The following code will pass the compilation, but fails with E0228, "the lifetime bound for this object type cannot be deduced from context; please supply an explicit bound" on the lambda with clippy:

```rust
trait Tr {}

fn f(_: u32) -> Option<Vec<Box<dyn Tr>>> {
    (0..1).map(|_| -> Option<Box<dyn Tr>> { todo!() }).collect()
}

fn main() {
    let _f0 = f;
    let _f0 = |_: u32| -> Option<Vec<Box<dyn Tr>>> {
        (0..1).map(|_| -> Option<Box<dyn Tr>> { todo!() }).collect()
    };
    _f0(20u32).unwrap();
}
```

Note that the `f` function implementation won't trigger such error, while the lambda defined in `main()` will trigger.

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bbbe62b834dfe344559744f1acd6d786).

### Reproducer

I tried this code:

```rust
trait Tr {}

fn f(_: u32) -> Option<Vec<Box<dyn Tr>>> {
    (0..1).map(|_| -> Option<Box<dyn Tr>> { todo!() }).collect()
}

fn main() {
    let _f0 = f;
    let _f0 = |_: u32| -> Option<Vec<Box<dyn Tr>>> {
        (0..1).map(|_| -> Option<Box<dyn Tr>> { todo!() }).collect()
    };
    _f0(20u32).unwrap();
}
```

I expected to see this happen:

Clippy will at least pass with warnings.

Instead, this happened:

Clippy fails with the following output:

```plaintext
Checking playground v0.0.1 (/playground)
error[[E0228]](https://doc.rust-lang.org/stable/error_codes/E0228.html): the lifetime bound for this object type cannot be deduced from context; please supply an explicit bound
 --> src/main.rs:9:42
  |
9 |     let _f0 = |_: u32| -> Option<Vec<Box<dyn Tr>>> {
  |                                          ^^^^^^

For more information about this error, try `rustc --explain E0228`.
error: could not compile `playground` (bin "playground") due to previous error
```

### Version

```text
`1.72.0`. I only tried with the stable channel on playground, so I can't run `rustc -Vv`.
```

**Comment by Alexendoo** [comment]
The cause seems to be

https://github.com/rust-lang/rust-clippy/blob/8c20739e4d9f1981203c516b8dc0c5288b704059/clippy_lints/src/types/vec_box.rs#L42

**Comment by lamalex** [comment]
Another example case: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=88b96f7fc298a6c643b933787ceb93dc
